src/js/controller/instream-adapter.js

```diff
@@ -89,6 +89,7 @@ const InstreamAdapter = function(_controller, _model, _view, _mediaPool) {
         _adProgram.setup();
 
         _adProgram.on('all', _instreamForward, this);
+        _adProgram.on('buffer', triggerAdBuffer, this);
```

Comment:
We already handle state changes in `_handleStateChange`.

Reply:
@robwalch Isn't this different from state event? (`buffering` vs `buffer`).
This `buffer` event comes from `program-listener` every time the buffer changes: https://github.com/jwplayer/jwplayer/blob/master/src/js/program/program-listeners.js#L73

Reply:
The point is to match the player state changes.

Player events | Ad mode events
------------ | -------------
buffer | adBuffer
play | adPlay
pause | adPause

Reply:
i'm confused by `handleStateChange`; the `newState` is coming from the provider, but the `oldState` is coming from the model. But the model's state is set to `buffering` before the provider state change , so `handleStateChange` detects a change from `buffering` to `loading` which is incorrect. Meanwhile the player's state is still `idle`. What should the flow be ?
https://github.com/jwplayer/jwplayer/pull/3169/files#diff-4b5d9c794240aa20cb620353eb027d56R114